### ISSUE TITLE
OPeNDAP inputs

### ISSUE THREAD

**Issue by rabernat** [pangeo-forge/pangeo-forge-recipes#54]
We will often want to pull data from OPeNDAP servers (see https://github.com/pangeo-forge/staged-recipes/issues/10).

But we have a choice to make. **Do we want to first cache the inputs from the OPeNDAP server into netCDF files?** Or do we want to stream directly from OPeNDAP to Zarr?

If we skip caching, this _almost_ works today. However, it will probably fail if we try it because we currently always try open the file using fsspec.

https://github.com/pangeo-forge/pangeo-forge/blob/f9b698a35a62ab169bce546b15fef9722c70cd90/pangeo_forge/recipe.py#L46-L50

So we need to refactor the recipe class to allow for more flexibility here.

Testing this will be tricky because we will have to spin up an opendap server in the test suite. There may be some examples in https://github.com/pydata/xarray/blob/master/xarray/tests/test_backends.py using pydap.

**Comment by charlesbluca** [comment]
For your first question, I suppose it comes down to what the process of OPeNDAP to Zarr looks like - is this essentially like streaming netCDF into memory and then writing that to Zarr? Or is it equivalent to downloading a Zarr dataset, with the conversion handled server-side?

If it's the former, then I think caching makes sense - like in the case of caching between different cloud providers to reduce egress, I think this would help to reduce likelihood of network failure and requests made to the server. If it's the latter, then I don't see that being as much of a concern.

As for testing, if all that is needed is some example netCDF files on an OPeNDAP server, it looks like xarray handles this by [accessing test data on a publicly available test server](https://github.com/pydata/xarray/blob/a0c71c1508f34345ad7eef244cdbbe224e031c1b/xarray/tests/test_backends.py#L3497-L3503):

```python
    def create_datasets(self, **kwargs):
        url = "http://test.opendap.org/opendap/hyrax/data/nc/bears.nc"
        actual = open_dataset(url, engine="pydap", **kwargs)
        with open_example_dataset("bears.nc") as expected:
            # workaround to restore string which is converted to byte
            expected["bears"] = expected["bears"].astype(str)
            yield actual, expected
```

**Comment by jsadler2** [comment]
Not sure if this is helpful or not. I wrote some code to pull data from an OPeNDAP server and write to a Zarr without caching. I had no problem skipping the caching to netcdf. I did this for pulling the NLDAS data from the NASA server.Here is a [link to the code ](https://github.com/jsadler2/preprocess_nwm_pgdl/blob/ea20d5465c180cb8362b8b47fdb069c4d37b8404/scripts/pull_nldas.py#L84)I wrote for this.

Regarding the "network failure" issue that you mention, @charlesbluca, the way I dealt with this is that I wrote code to delete the last chunk of my Zarr on the time dimension (that was my `append` dimension) and then to start pulling from the OPeNDAP server from that last chunk on. This way, if there was a network failure, it would basically start adding to the Zarr dataset from where it left off. This functionality also serves as a way to update the Zarr dataset when new data (in the time dimension) are addded.
